backend/wmg/api/rollup.py

```diff
@@ -0,0 +1,130 @@
+import owlready2
+import pandas as pd
+import numpy as np
+import numba as nb
+from functools import lru_cache
+from backend.wmg.data.constants import CL_BASIC_PERMANENT_URL_OWL
+
+
+# ontology object
+onto = None
+
+
+def find_descendants_per_cell_type(cell_types):
+    """
+    Find descendants for each cell type in the input list.
+
+    Parameters
+    ----------
+    cell_types : list
+        List of cell types (cell type ontology term IDs) to find descendants for
+
+    Returns
+    -------
+    descendants_per_cell_type : list
+        List of lists of descendants for each cell type in the input list.
+    """
+
+    global onto
+    if not onto:
+        onto = owlready2.get_ontology(CL_BASIC_PERMANENT_URL_OWL)
+        onto.load()
+
+    # cache finding descendants per cell type
+    @lru_cache(maxsize=None)
+    def _descendants(cell_type):
+        cell_type_iri = cell_type.replace(":", "_")
+        entity = onto.search_one(iri=f"http://purl.obolibrary.org/obo/{cell_type_iri}")
+        if entity:
+            descendants = [i.name.replace("_", ":") for i in entity.descendants()]
+        else:
+            descendants = [cell_type]
+        return descendants
+
+    descendants_per_cell_type = []
+    for cell_type in cell_types:
+        descendants = _descendants(cell_type)
+        descendants_per_cell_type.append(descendants)
+
+    for i, children in enumerate(descendants_per_cell_type):
+        # remove descendent cell types that are not cell types in the WMG data
+        descendants_per_cell_type[i] = list(set(children).intersection(cell_types))
+
+    return descendants_per_cell_type
+
+
+def rollup_across_cell_type_descendants(df, cell_type_col="cell_type_ontology_term_id"):
```

Comment:
This module was relocated, so the diff is hard to parse. The main update was to rewrite this function to accept a tidy dataframe with `N` non-numeric and `M` numeric columns. The `N` non-numeric columns must contain cell type ontology term IDs, and are treated as the dimensions of an multi-dimensional numpy array. The numeric data is slotted into this array and rolled up along the first axis (which will always correspond to the cell type ontology term IDs). The resulting rolled up array is reshaped back into the tidy dataframe and returned.

This ensures that cell types are only rolled up within each combination of other dimensions (e.g. tissue, gene, organism).

Reply:
most of this is useful as a comment in the source (from `The non-numeric columns must contain cell type ontology term IDs,...` on)